### ISSUE TITLE
private CA cert, CA chain, Two-Way SSL

### ISSUE THREAD

**Issue by eulmatxlm** [google/grr#735]
Hello,

We're working with GRR v3.3.0.4, installed on Ubuntu from the release debs.

And we are facing issues when dealing with some requirements during our deployments:

- First, we need to be able to use a private CA certificate (a chain actually) for HTTPS communication between the clients and the load-balancer/reverse proxy
- We also need client certificate authentication

For the custom CA, short of opening the zip templates to replace the cacert.pem bundled with certifi, it does not seem to be possible without modifying the code of the client builder.
Same issue with client certificate authentication, which is not implemented.

Are we missing something there?

Would be great to have a confirmation that we actually need to produce some code before starting the heavy lifting. Especially as we can already foresee some difficulties:

-  with urllib/requests which support only path to files for client certificate authentication (so no trivial way to put that in the configuration)
- and the config module itself which, unless we're mistaken, will probably choke if we try to have more than one certificate in a single configuration entry, a certificate chain being usually the concatenation of a root CA and its (multiple) subCA ...

Thanks for your help!

**Comment by bgalehouse** [comment]
What is your timeframe? Internally, we are now using [Fleetspeak](https://github.com/google/fleetspeak) for comms with most clients. We are still working on setting up packaging and documentation for our open-source users, but the code is there and operational artifacts are under active development.

I mention this both because FS may be closer to what you are looking for, more or less out of the box, and because any serious setup work on this front is likely to need to be redone with FS.

One thing to be clear of is that the current GRR clients do not use HTTPS, instead they use HTTP with payloads armored in a manner roughly equivalent to an older HTTPS standard - in particular, there is no perfect forward secrecy, and cypher choices are limited.

FS, on the other hand, uses the Go language's native HTTPS, coded to be restricted to more modern cryptographic modes. The FS client code is structured make it possible to compile clients with trusted roots hardcoded in the client binary. We do this internally, so that the root CA cert and server endpoint are protected by any binary level signatures and hash-based whitelisting. This is something that we want others to be able to do, but I don't yet know how perfectly this will be supported by the general packaging and deployment strategy that we end up with.

Also, neither FS and GRR really support client certificates in the usual sense. In the case of FS you could add it by tweaking/wrapping the HTTPS communicator component on the server. However, the reason that we don't do client certs is because we believe that clients should be zero conf.

Basically, the FS client presents a self-signed cert and the server remembers a hash of the client's public key to use as the client's ID. This means that clients can essentially always talk to the server, but cannot impersonate another client. We think this is important, so that analysts can investigate clients that are in some sense only partially setup or otherwise broken. On the other hand, it is trivial to pull the ID (key fingerprint) that a client is using out of the FS logfile. So one option would be to check that ID after starting FS for the first time on a new install, and then add some sort of 'trusted' label to that ID on the GRR server. Then you could easily restrict hunts to those clients which successfully went through this full process.

**Comment by eulmatxlm** [comment]
Hughh, and here we were just starting to get used to all the GRR components...

Just had a quick look at FS, and your remark about the timeframe actually will dictate our next steps. The first building steps went ok apart from a dependency issue (grpcio-tools==1.17.1), however the installation and integration might take too much time for now.

More to the point, what we want is client authentication, be it via certificates or any other means.  We'll try to evaluate FS further, but in parallel we'll try to write a quick patch for the features we need in the "regular" GRR. As said previously, we're kind of comfortable with GRR code, it will be easier/faster than to dig into FS to meet our requirements.

As a point of notice, we also use full encryption with client certificate for the connection to the database, which seems to not be implemented right now in FS...

**Comment by bgalehouse** [comment]
Regarding the database question, we use the Go-MySQL-Driver, and by the [docs](https://github.com/go-sql-driver/mysql#tls), you can enable TLS in the connection string, and configure custom TLS parameters (to add, e.g. a client cert) with the addition of a single call. I don't think there would be pushback against adding client cert functionality to the general-purpose FS server binary.

Regarding endpoint authentication, both FS and classic GRR have an identity model (the FS identity model reused the concept of the GRR identity model) which is secure against impersonation, but allows hosts to create new identities as needed. Because there is protection against impersonation, this can be built up to provide stronger guarantees.

FS also provides a mechanism (by adding code) for the client to sign comms with an additional cert, and for the server to see this before deciding whether to accept communication from the client.

On the topic of identity, what is your desired behavior is if a machine or VM is duplicated? With FS there is functionality available to detect duplicate machines and the force all machines using the same identity to rekey. This is an example of what we've find problematic with classic client cert authentication systems.

**Comment by eulmatxlm** [comment]
Yes, actually just built and run the test.sh successfully by modifying mysql_test.go with the [RegisterTLSConfig](https://godoc.org/github.com/go-sql-driver/mysql#RegisterTLSConfig) example, so it should not be that complicated.

Again, the issue here is the amount of time we can dedicate to FS. GRR components are well understood on our side, whereas we have no experience at all with FS. And very little with the go programing language , for that matter. By any chance, do you have a short documentation at hand explaining how to deploy FS with GRR?

Regarding identity, when we deploy GRR, it's only for a short period of time on a known set of workstations, therefore duplication is kind of a non-issue for us. Although it's easy to imagine the benefit of having such controls to be able to force client's rekeying. What we do want, however, is to restrict access to the frontend to authenticated machines. And we obviously can't filter just on IPs, as the clients are often/always behind a NAT. Hence the client certificate authentication, which cuts short any connection attempts from unknown clients.

**Comment by mbushkov** [comment]
@eulmatxlm , we don't have a documentation explaining FS+GRR deployment just yet. We're working on providing an easy way to integrate FS with GRR right now and  plan to have this work done (and thus - documentation written) this quarter.

Answering your questions from the original post:
> with urllib/requests which support only path to files for client certificate authentication (so no trivial way to put that in the configuration)

If you're ready to modify the client, you can still store certificates in the configuration, but for example, put them in a temp file on client startup. This would allow using them with urllib/requests.

> and the config module itself which, unless we're mistaken, will probably choke if we try to have more than one certificate in a single configuration entry, a certificate chain being usually the concatenation of a root CA and its (multiple) subCA ...

You could probably define a new configuration option usinge config_lib.DEFINE_list (like it's done [here](https://github.com/google/grr/blob/b719185ad29ad8aa7b5b156b3954366a30c71b83/grr/core/grr_response_core/config/artifacts.py#L10)), store the certificates in hex-encoded form as strings and then use this option in the custom code you'll add to the client.
